File: zul/src/main/java/org/zkoss/zul/Codeeditor.java
/* Codeeditor.java

        Purpose:
                
        Description:
                
        History:
                Thu Jun 18 18:05:45 CST 2026, Created by peakerlee

Copyright (C) 2026 Potix Corporation. All Rights Reserved.
*/
package org.zkoss.zul;

import java.io.IOException;

import org.zkoss.lang.Objects;
import org.zkoss.zk.au.AuRequest;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.event.InputEvent;
import org.zkoss.zk.ui.ext.Disable;
import org.zkoss.zk.ui.ext.Readonly;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.impl.Utils;
import org.zkoss.zul.impl.XulElement;

/**
 * A code editor that loads and edits source code with syntax highlighting.
 *
 * <p>The Community Edition build provides basic editing: setting/loading the
 * content ({@link #setValue}), choosing the highlight language
 * ({@link #setLanguage}), toggling read-only ({@link #isReadonly}) and line
 * numbers ({@link #setLineNumbers}), and an optional light/dark editor theme
 * ({@link #setTheme}). Edits are reported to the server through the standard
 * {@code onChange}/{@code onChanging} events carrying an {@link InputEvent}, so
 * {@code value} participates in data binding the same way as other input
 * components.
 *
 * <p>Advanced capabilities (autocomplete, lint, code folding, multi-cursor,
 * search/replace, diff) are not part of the CE build; they are layered on by
 * the EE module through a client-side extension hook, so an application can
 * adopt them later without changing the component or the ZUL.
 *
 * @author peakerlee
 * @since 10.4.0
 */
public class Codeeditor extends XulElement implements Disable, Readonly {
	/** The accepted {@link #setLanguage} tokens. The default {@code "plain"}
	 * must be present, otherwise explicitly setting {@code language="plain"}
	 * would be rejected by {@link Utils#checkEnum}.
	 */
	private static final String[] LANGS = { "plain", "html", "xml", "java", "javascript", "css", "json", "sql",
			"markdown" };

	private String _value = "";
	private String _language = "plain";
	private boolean _readonly;
	private boolean _disabled;
	private boolean _lineNumbers = true;
	/** Unset by default: when null the client derives the editor theme from
	 * {@code zk.themeName}; only an explicitly set value is rendered. */
	private String _theme;

	static {
		addClientEvent(Codeeditor.class, Events.ON_CHANGE, CE_IMPORTANT | CE_REPEAT_IGNORE);
		addClientEvent(Codeeditor.class, Events.ON_CHANGING, CE_DUPLICATE_IGNORE);
		addClientEvent(Codeeditor.class, Events.ON_FOCUS, CE_DUPLICATE_IGNORE);
		addClientEvent(Codeeditor.class, Events.ON_BLUR, CE_DUPLICATE_IGNORE);
	}

	public Codeeditor() {
	}

	public Codeeditor(String value) {
		setValue(value);
	}

	/**
	 * Returns the editor content.
	 * <p>Default: an empty string.
	 */
	public String getValue() {
		return _value;
	}

	/**
	 * Sets the editor content.
	 *
	 * @param value the source code; {@code null} is treated as an empty string.
	 * @since 10.4.0
	 */
	public void setValue(String value) {
		if (value == null)
			value = "";
		if (!Objects.equals(_value, value)) {
			_value = value;
			smartUpdate("value", _value);
		}
	}

	/**
	 * Returns the syntax-highlight language.
	 * <p>Default: {@code "plain"}.
	 */
	public String getLanguage() {
		return _language;
	}

	/**
	 * Sets the syntax-highlight language.
	 *
	 * @param language one of {@code plain}, {@code html}, {@code xml},
	 * {@code java}, {@code javascript}, {@code css}, {@code json}, {@code sql},
	 * {@code markdown}; {@code null} resets it to the default {@code "plain"}.
	 * @exception WrongValueException if {@code language} is non-null and not an
	 * accepted token.
	 * @since 10.4.0
	 */
	public void setLanguage(String language) {
		language = Utils.checkEnum(language, "plain", "Unknown language: ", LANGS);
		if (!Objects.equals(_language, language)) {
			_language = language;
			smartUpdate("language", _language);
		}
	}

	/**
	 * Returns whether the editor is read-only.
	 * <p>Default: false.
	 */
	public boolean isReadonly() {
		return _readonly;
	}

	/**
	 * Sets whether the editor is read-only.
	 *
	 * @param readonly whether to forbid editing the content.
	 * @since 10.4.0
	 */
	public void setReadonly(boolean readonly) {
		if (_readonly != readonly) {
			_readonly = readonly;
			smartUpdate("readonly", _readonly);
		}
	}

	/**
	 * Returns whether the editor is disabled.
	 * <p>Default: false.
	 */
	public boolean isDisabled() {
		return _disabled;
	}

	/**
	 * Sets whether the editor is disabled. A disabled editor cannot be edited
	 * or focused and is shown dimmed; unlike {@link #setReadonly}, which keeps
	 * the content focusable and selectable.
	 *
	 * @param disabled whether to disable the editor.
	 * @since 10.4.0
	 */
	public void setDisabled(boolean disabled) {
		if (_disabled != disabled) {
			_disabled = disabled;
			smartUpdate("disabled", _disabled);
		}
	}

	/**
	 * Returns whether the gutter line numbers are shown.
	 * <p>Default: true.
	 */
	public boolean isLineNumbers() {
		return _lineNumbers;
	}

	/**
	 * Sets whether the gutter line numbers are shown.
	 *
	 * @param lineNumbers whether to show the line-number gutter.
	 * @since 10.4.0
	 */
	public void setLineNumbers(boolean lineNumbers) {
		if (_lineNumbers != lineNumbers) {
			_lineNumbers = lineNumbers;
			smartUpdate("lineNumbers", _lineNumbers);
		}
	}

	/**
	 * Returns the editor theme, or {@code null} when unset.
	 * <p>Default: {@code null}. When unset the client derives the theme from
	 * the page's {@code zk.themeName} (a dark ZK theme yields the dark editor
	 * theme, otherwise light), so the editor follows the page without the
	 * server pushing a value the client already knows.
	 */
	public String getTheme() {
		return _theme;
	}

	/**
	 * Sets the editor theme to explicitly override the {@code zk.themeName}
	 * derived default.
	 *
	 * @param theme {@code light} or {@code dark}; {@code null} restores the
	 * client-derived default.
	 * @exception WrongValueException if {@code theme} is non-null and neither
	 * {@code light} nor {@code dark}.
	 * @since 10.4.0
	 */
	public void setTheme(String theme) {
		if (theme != null)
			theme = Utils.checkEnum(theme, null, "Unknown theme: ", "light", "dark");
		if (!Objects.equals(_theme, theme)) {
			_theme = theme;
			smartUpdate("theme", _theme);
		}
	}

	protected boolean isChildable() {
		return false;
	}

	protected void renderProperties(ContentRenderer renderer) throws IOException {
		super.renderProperties(renderer);
		if (!_value.isEmpty())
			render(renderer, "value", _value);
		if (!"plain".equals(_language))
			render(renderer, "language", _language);
		if (_readonly)
			render(renderer, "readonly", true);
		if (_disabled)
			render(renderer, "disabled", true);
		if (!_lineNumbers)
			// lineNumbers defaults to true, so the non-default is false; the
			// boolean render(...) helper skips false, so send it explicitly.
			renderer.render("lineNumbers", false);
		if (_theme != null)
			render(renderer, "theme", _theme);
	}

	//-- ComponentCtrl --//
	public void service(AuRequest request, boolean everError) {
		final String cmd = request.getCommand();
		if (cmd.equals(Events.ON_CHANGE) || cmd.equals(Events.ON_CHANGING)) {
			// Rebuild the payload through the factory (reads only data.get("value"));
			// never trust request.getData() wholesale.
			final InputEvent evt = InputEvent.getInputEvent(request, _value);
			if (cmd.equals(Events.ON_CHANGE)) {
				if (Objects.equals(_value, evt.getValue()))
					return; //Bug 1881557: don't post onChange when the value wasn't modified
				_value = evt.getValue(); // keep the server-side value in sync on commit
			}
			Events.postEvent(evt);
		} else
			super.service(request, everError);
	}
}

File: zul/src/main/resources/web/js/zul/code/Codeeditor.ts
/* Codeeditor.ts

	Purpose:

	Description:

	History:
		Wed Jun 17 22:25:01 CST 2026, Created by peakerlee

Copyright (C) 2026 Potix Corporation. All Rights Reserved.
*/
import { EditorState, Compartment, Prec, type Extension } from '@codemirror/state';
import { EditorView, lineNumbers, keymap } from '@codemirror/view';
import { defaultKeymap, history, historyKeymap, indentWithTab } from '@codemirror/commands';
import { syntaxHighlighting, defaultHighlightStyle, HighlightStyle, indentUnit } from '@codemirror/language';
import { tags } from '@lezer/highlight'; // CodeMirror's highlight-tag vocabulary; a hard peer of @codemirror/language
import { xml } from '@codemirror/lang-xml';
import { java } from '@codemirror/lang-java';
import { javascript } from '@codemirror/lang-javascript';
import { css } from '@codemirror/lang-css';
import { json } from '@codemirror/lang-json';
import { html } from '@codemirror/lang-html';
import { sql } from '@codemirror/lang-sql';
import { markdown } from '@codemirror/lang-markdown';

/**
 * Dark-surface syntax palette, used when the editor renders dark; light mode
 * keeps CodeMirror's `defaultHighlightStyle`. Colors are VS Code Dark+ inspired
 * to read well on the `#1e1e1e` surface, and each is themeable through a
 * `--zk-codeeditor-token-*` custom property (see the component LESS), with the
 * literal as the fallback when no theme overrides it.
 */
const darkHighlightStyle = HighlightStyle.define([
	{ tag: tags.keyword, color: 'var(--zk-codeeditor-token-keyword, #569cd6)' },
	{ tag: [tags.string, tags.special(tags.string)], color: 'var(--zk-codeeditor-token-string, #ce9178)' },
	{ tag: [tags.comment, tags.lineComment, tags.blockComment], color: 'var(--zk-codeeditor-token-comment, #6a9955)', fontStyle: 'italic' },
	{ tag: [tags.number, tags.bool, tags.null, tags.atom], color: 'var(--zk-codeeditor-token-number, #b5cea8)' },
	{ tag: [tags.function(tags.variableName), tags.function(tags.propertyName)], color: 'var(--zk-codeeditor-token-function, #dcdcaa)' },
	{ tag: [tags.typeName, tags.className, tags.namespace], color: 'var(--zk-codeeditor-token-type, #4ec9b0)' },
	{ tag: [tags.variableName, tags.propertyName], color: 'var(--zk-codeeditor-token-variable, #9cdcfe)' },
	{ tag: tags.tagName, color: 'var(--zk-codeeditor-token-tag, #569cd6)' },
	{ tag: tags.attributeName, color: 'var(--zk-codeeditor-token-attribute, #9cdcfe)' },
	{ tag: [tags.meta, tags.regexp, tags.escape], color: 'var(--zk-codeeditor-token-meta, #d16969)' },
]);

/**
 * A code editor that loads and edits source code with syntax highlighting,
 * the client half of the server-side `org.zkoss.zul.Codeeditor` component.
 *
 * <p>The Community Edition build wraps a self-assembled CodeMirror 6 editor
 * (line numbers, history, syntax highlighting, editable toggle) and reports
 * edits to the server as `onChanging` (while typing) and `onChange` (on blur).
 * The grammars for all supported languages are bundled into this package, so
 * only pages that actually use `<codeeditor>` pay for the editor weight.
 *
 * @author peakerlee
 * @since 10.4.0
 */
@zk.WrapClass('zul.code.Codeeditor')
export class Codeeditor extends zul.Widget {
	/** @internal */
	_value = '';
	/** @internal */
	_language = 'plain';
	/** @internal */
	_readonly = false;
	/** @internal */
	_disabled = false;
	/** @internal */
	_lineNumbers = true;
	/** @internal unset → derived from `zk.themeName` (a dark page theme yields the dark editor). */
	_theme?: string;
	/** @internal */
	_cm?: EditorView;
	/** @internal value last synchronized with the server (initial, after a commit, or after a programmatic set); the onChange/blur commit baseline. */
	_committed = '';
	/** @internal true while `_setDoc` pushes a server/programmatic value, so the doc-change listener does not echo it back to the server as `onChanging`. */
	_settingDoc = false;
	/** @internal page CSP nonce, captured once in `bind_` before the bootstrap scripts are pruned, then reused across rerenders. */
	_nonce?: string;
	/** @internal */
	_langComp = new Compartment();
	/** @internal */
	_roComp = new Compartment();
	/** @internal */
	_lnComp = new Compartment();
	/** @internal */
	_themeComp = new Compartment();
	/** @internal toggles `indentWithTab` in/out — see `_toggleTabIndent`. */
	_tabComp = new Compartment();
	/** @internal false (default) → Tab moves focus (no keyboard trap); true → Tab indents. Toggled by Escape. */
	_tabIndent = false;
	/** @internal reconfigures the tab display width — see `setTabSize`. */
	_tabSizeComp = new Compartment();
	/** @internal columns a tab character is rendered as (CM `EditorState.tabSize`). */
	_tabSize = 4;

	/**
	 * Returns the editor content.
	 */
	getValue(): string {
		return this._value;
	}

	/**
	 * Sets the editor content.
	 * @param value - the source code.
	 * @param opts - pass `{force: true}` to push the value even when unchanged.
	 */
	setValue(value: string, opts?: Record<string, boolean>): this {
		const o = this._value;
		this._value = value ?? '';
		if (this._cm && (o !== this._value || opts?.force))
			this._setDoc(this._value);
		return this;
	}

	/**
	 * Returns the syntax-highlight language.
	 */
	getLanguage(): string {
		return this._language;
	}

	/**
	 * Sets the syntax-highlight language and reconfigures the live editor.
	 * @param language - one of the supported grammar tokens, or `plain`.
	 */
	setLanguage(language: string, opts?: Record<string, boolean>): this {
		const o = this._language;
		this._language = language || 'plain';
		if (this._cm && (o !== this._language || opts?.force))
			this._cm.dispatch({ effects: this._langComp.reconfigure(this._langExtension()) });
		return this;
	}

	/**
	 * Returns whether the editor is read-only.
	 */
	isReadonly(): boolean {
		return this._readonly;
	}

	/**
	 * Sets whether the editor is read-only and reconfigures the live editor.
	 * @param readonly - whether to forbid editing.
	 */
	setReadonly(readonly: boolean, opts?: Record<string, boolean>): this {
		const o = this._readonly;
		this._readonly = readonly;
		if (this._cm && (o !== readonly || opts?.force))
			this._applyEditable();
		return this;
	}

	/**
	 * Returns whether the editor is disabled.
	 */
	isDisabled(): boolean {
		return this._disabled;
	}

	/**
	 * Sets whether the editor is disabled: not editable, not focusable, dimmed.
	 * @param disabled - whether to disable the editor.
	 */
	setDisabled(disabled: boolean, opts?: Record<string, boolean>): this {
		const o = this._disabled;
		this._disabled = disabled;
		if (this._cm && (o !== disabled || opts?.force)) {
			this._applyEditable();
			this._applyDisabledClass();
		}
		return this;
	}

	/** @internal reconfigures the live editor's editable/read-only state from `_readonly || _disabled`. */
	_applyEditable(): void {
		this._cm!.dispatch({ effects: this._roComp.reconfigure(this._editableExtension()) });
	}

	/**
	 * Returns whether the gutter line numbers are shown.
	 */
	isLineNumbers(): boolean {
		return this._lineNumbers;
	}

	/**
	 * Sets whether the gutter line numbers are shown and reconfigures the editor.
	 * @param lineNumbers - whether to show the line-number gutter.
	 */
	setLineNumbers(lineNumbers: boolean, opts?: Record<string, boolean>): this {
		const o = this._lineNumbers;
		this._lineNumbers = lineNumbers;
		if (this._cm && (o !== lineNumbers || opts?.force))
			this._cm.dispatch({ effects: this._lnComp.reconfigure(this._lineNumbersExtension()) });
		return this;
	}

	/**
	 * Returns the number of columns a tab character is rendered as.
	 */
	getTabSize(): number {
		return this._tabSize;
	}

	/**
	 * Sets how many columns a tab character is rendered as and reconfigures the editor.
	 * @param tabSize - a positive column count (default 4); the indent character itself stays a tab.
	 */
	setTabSize(tabSize: number, opts?: Record<string, boolean>): this {
		const o = this._tabSize;
		this._tabSize = tabSize;
		if (this._cm && (o !== tabSize || opts?.force))
			this._cm.dispatch({ effects: this._tabSizeComp.reconfigure(this._tabSizeExtension()) });
		return this;
	}

	/**
	 * Returns the explicitly set editor theme, or `undefined` when it follows
	 * the page's `zk.themeName`.
	 */
	getTheme(): string | undefined {
		return this._theme;
	}

	/**
	 * Sets the editor theme to override the `zk.themeName` derived default.
	 * @param theme - `light`, `dark`, or `undefined` to follow the page.
	 */
	setTheme(theme: string, opts?: Record<string, boolean>): this {
		const o = this._theme;
		this._theme = theme;
		if (this._cm && (o !== theme || opts?.force)) {
			this._cm.dispatch({ effects: this._themeComp.reconfigure(this._themeExtension()) });
			this._applyDarkClass();
		}
		return this;
	}

	/** @internal */
	override bind_(desktop?: zk.Desktop, skipper?: zk.Skipper, after?: CallableFunction[]): void {
		super.bind_(desktop, skipper, after);
		// Read once and keep it: the page nonce is constant, and the bootstrap
		// scripts (where it is read from) are pruned during afterMount, so a later
		// rerender's bind_ can no longer find it — reuse the first good read.
		if (this._nonce === undefined)
			this._nonce = this._cspNonce();
		zk.afterMount(() => this._mount()); // CM needs a stable DOM node
	}

	/** @internal */
	override unbind_(skipper?: zk.Skipper, after?: CallableFunction[], keepRod?: boolean): void {
		if (this._cm) {
			// null _cm BEFORE destroy: CM's destroy() blurs a focused contentDOM,
			// firing a synchronous focusout — the handler's `if (!this._cm)` guard
			// must already see undefined so it doesn't commit/fire onBlur mid-unbind.
			const cm = this._cm;
			this._cm = undefined;
			cm.destroy();
		}
		super.unbind_(skipper, after, keepRod);
	}

	/** @internal builds the fixed CE editor set; EE layers more via `_extraExtensions`. */
	_baseExtensions(): Extension[] {
		return [
			history(),
			indentUnit.of('\t'), // indent unit = a tab character (Tab-indent and Ctrl-]/Ctrl-[)
			keymap.of([...defaultKeymap, ...historyKeymap]),
			// Escape toggles "editing mode": Tab moves focus by default (no keyboard
			// trap), indents after Escape, and returns to focus-nav on Escape again.
			// Prec.low so a future EE popup (autocomplete/search) consumes Escape first.
			Prec.low(keymap.of([{ key: 'Escape', run: (): boolean => { this._toggleTabIndent(); return true; } }])),
			EditorView.updateListener.of((u) => {
				if (u.docChanged)
					this._onDocChanged(u.state.doc.toString());
			}),
			this._lnComp.of(this._lineNumbersExtension()),
			this._roComp.of(this._editableExtension()),
			this._langComp.of(this._langExtension()),
			this._themeComp.of(this._themeExtension()),
			this._tabComp.of(this._tabExtension()),
			this._tabSizeComp.of(this._tabSizeExtension()),
			...this._extraExtensions(),
		];
	}

	/**
	 * @internal Sanctioned extension point for CodeMirror extensions contributed
	 * by subclasses (e.g. a markdown editor's bold/italic keymap) and by the EE
	 * build (autocomplete/lint/fold/…). CE itself contributes none.
	 *
	 * Contract: an override MUST chain — `return super._extraExtensions().concat(...)` —
	 * and never replace without calling super, so that a subclass's extensions and
	 * the EE augmentation compose instead of shadowing one another.
	 */
	_extraExtensions(): Extension[] {
		return [];
	}

	/** @internal */
	_mount(): void {
		if (this._cm || !this.desktop)
			return;
		const cave = this.$n('cave');
		if (!cave)
			return;
		const exts = this._baseExtensions();
		if (this._nonce)
			exts.push(EditorView.cspNonce.of(this._nonce)); // CSP strict-dynamic: tag CM's injected <style>
		// forward an app-supplied accessible name (ca:aria-label / ca:aria-labelledby on
		// the root) onto CM's editable textbox, where screen readers actually need it.
		const root = this.$n(),
			// always describe the Tab/Escape keyboard model; add an app-supplied name when present.
			aria: Record<string, string> = { 'aria-describedby': this.uuid + '-status' },
			label = root?.getAttribute('aria-label'),
			labelledby = root?.getAttribute('aria-labelledby');
		if (label) aria['aria-label'] = label;
		if (labelledby) aria['aria-labelledby'] = labelledby;
		exts.push(EditorView.contentAttributes.of(aria));
		this._cm = new EditorView({
			state: EditorState.create({ doc: this._value, extensions: exts }),
			parent: cave,
		});
		// CodeMirror normalizes line endings (CRLF/CR → LF) in its doc. Baseline
		// _committed/_value from the actual doc, not the raw server string, so a
		// CRLF value doesn't look "changed" on the first blur and fire a spurious
		// onChange that rewrites the server value to LF.
		this._committed = this._value = this._cm.state.doc.toString();
		this._applyDarkClass(); // surface theming is applied via the z-codeeditor-dark LESS class
		this._applyDisabledClass();
		const dom = this._cm.contentDOM;
		// CM6 has no native blur-commit: commit onChange when the editor loses focus.
		// Compare against _committed, not _value — _value tracks every onChanging
		// keystroke, so guarding on it would suppress the commit entirely.
		dom.addEventListener('focusout', () => {
			if (!this._cm)
				return; // editor was destroyed (detach/unbind) while focused
			const v = this._cm.state.doc.toString();
			if (v !== this._committed) {
				this._committed = this._value = v;
				this.fire('onChange', { value: v });
			}
			this.fire('onBlur');
		});
		dom.addEventListener('focusin', () => this.fire('onFocus'));
		this._onMounted();
	}

	/**
	 * @internal Called at the end of `_mount`, only after the CodeMirror instance
	 * has been created and bound. Subclasses override this — not `_mount` — to wire
	 * post-mount behavior (a toolbar, an initial preview, …) without having to guard
	 * against `_mount`'s early returns.
	 */
	_onMounted(): void {
		// no-op in CE; a sanctioned post-mount extension point for subclasses.
	}

	/** @internal pushes a new document into the live editor, replacing its content. */
	_setDoc(value: string): void {
		const cm = this._cm!,
			cur = cm.state.doc.toString();
		if (cur !== value) {
			this._settingDoc = true; // this is a server/programmatic write, not a user edit
			try {
				cm.dispatch({ changes: { from: 0, to: cur.length, insert: value } });
			} finally {
				this._settingDoc = false;
			}
		}
		this._committed = value; // a programmatic set is authoritative — no blur commit needed
	}

	/** @internal high-frequency keystroke notification; AU queue may drop stale ones. */
	_onDocChanged(value: string): void {
		this._value = value;
		// A server/programmatic `_setDoc` also fires CodeMirror's doc-change listener;
		// never echo that back as if the user typed it. And, like InputWidget, only
		// notify the server when something listens (the binder saves on onChange).
		if (this._settingDoc || !this.isListen('onChanging'))
			return;
		this.fire('onChanging', { value: value }, { ignorable: true, rtags: { onChanging: 1 } });
	}

	/** @internal */
	_langExtension(): Extension {
		switch (this._language) {
		case 'xml': return xml();
		case 'java': return java();
		case 'javascript': return javascript();
		case 'css': return css();
		case 'json': return json();
		case 'html': return html();
		case 'sql': return sql();
		case 'markdown': return markdown();
		default: return []; // plain
		}
	}

	/** @internal editing is blocked when either read-only or disabled. */
	_editableExtension(): Extension {
		const locked = this._readonly || this._disabled;
		return [EditorView.editable.of(!locked), EditorState.readOnly.of(locked)];
	}

	/** @internal `indentWithTab` only in editing mode; otherwise Tab is unbound so it moves focus. */
	_tabExtension(): Extension {
		return this._tabIndent ? keymap.of([indentWithTab]) : [];
	}

	/**
	 * @internal Flips Tab between focus navigation (default) and 4-space indentation,
	 * driven by Escape. Default focus-nav means the editor is never a keyboard trap
	 * (WCAG 2.1.2); the user opts into indentation with Escape and leaves with Escape
	 * again. The change is announced through the `aria-live` status node.
	 */
	_toggleTabIndent(): void {
		this._tabIndent = !this._tabIndent;
		this._cm?.dispatch({ effects: this._tabComp.reconfigure(this._tabExtension()) });
		this._announceTabMode();
	}

	/** @internal updates the visually-hidden `aria-live` region so the current Tab behavior is announced. */
	_announceTabMode(): void {
		const status = this.$n('status');
		if (status)
			status.textContent = this._tabIndent
				? 'Tab key now indents. Press Escape to return to focus navigation.'
				: 'Tab key now moves focus. Press Escape to enable indentation.';
	}

	/** @internal */
	_lineNumbersExtension(): Extension {
		return this._lineNumbers ? lineNumbers() : [];
	}

	/** @internal renders a tab character as `_tabSize` columns wide. */
	_tabSizeExtension(): Extension {
		return EditorState.tabSize.of(this._tabSize);
	}

	/** @internal whether the editor uses its dark surface: explicit `theme="dark"`, else derived from a dark page theme. */
	_isDark(): boolean {
		return this._theme ? this._theme === 'dark' : !!zk.themeName?.includes('dark');
	}

	/**
	 * @internal Selects the syntax-highlight palette for the current surface, and
	 * in dark mode also tells CodeMirror it is dark so its built-in extensions
	 * (selection, matching-bracket…) pick their dark variants. Lives in the theme
	 * compartment so a live `setTheme` recolors tokens, not just the chrome. The
	 * surface chrome (background/gutter/cursor) stays in the component LESS
	 * (`.z-codeeditor-dark`); only the token palette is chosen here.
	 */
	_themeExtension(): Extension {
		return this._isDark()
			? [EditorView.theme({}, { dark: true }), syntaxHighlighting(darkHighlightStyle)]
			: syntaxHighlighting(defaultHighlightStyle);
	}

	/** @internal toggles the `z-codeeditor-dark` modifier so the LESS surface rules apply. */
	_applyDarkClass(): void {
		this._toggleClass('dark', this._isDark());
	}

	/** @internal toggles a `z-codeeditor-<name>` modifier class on the root, when bound. */
	_toggleClass(name: string, on: boolean): void {
		this.$n()?.classList.toggle(this.$s(name), on);
	}

	/** @internal toggles the `z-codeeditor-disabled` modifier (dim + non-interactive in LESS). */
	_applyDisabledClass(): void {
		this._toggleClass('disabled', this._disabled);
	}

	/** @internal reads the page nonce from a still-present nonced node for CM's StyleModule. */
	_cspNonce(): string | undefined {
		const n = document.querySelector('script[nonce]') ?? document.querySelector('style[nonce]');
		return (n as unknown as { nonce?: string })?.nonce || undefined;
	}
}
